File: pipeline/normalize/terms.py
"""Canonical terms encoding + terms_hash. Must match internal/schema/terms.go byte-for-byte."""

from __future__ import annotations

import hashlib
import json
from typing import Mapping

# Fixed field order — MUST NOT CHANGE without a schema_version bump.
_FIELD_ORDER: tuple[str, ...] = (
    "commitment",
    "tenancy",
    "os",
    "support_tier",
    "upfront",
    "payment_option",
)


def canonicalize_terms(terms: Mapping[str, str]) -> str:
    """Return the canonical JSON encoding of a terms mapping.

    Missing keys raise KeyError — callers are expected to fill defaults
    (via terms_defaults.yaml) before hashing.
    """
    values = [terms[k] for k in _FIELD_ORDER]
    # separators=(",", ":") -> no whitespace; ensure_ascii=False -> UTF-8 passthrough.
    return json.dumps(values, separators=(",", ":"), ensure_ascii=False)


def terms_hash(terms: Mapping[str, str]) -> str:
    """128-bit hex digest of the canonical encoding (first 32 hex chars of sha256)."""
    canonical = canonicalize_terms(terms)
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()[:32]

File: internal/schema/terms.go
// Package schema holds shared data-schema primitives used by both the catalog
// reader and (eventually) code-generated clients. The canonical terms encoding
// in this file MUST remain byte-identical to pipeline/normalize/terms.py —
// their equality is asserted by testdata/terms_golden.jsonl which is consumed
// by both test suites.
package schema

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
)

// Terms is the six-tuple that participates in terms_hash.
// Field order in the JSON array is fixed (see CanonicalizeTerms) and a
// schema_version bump is required to change it.
type Terms struct {
	Commitment    string
	Tenancy       string
	OS            string
	SupportTier   string
	Upfront       string
	PaymentOption string
}

// CanonicalizeTerms returns the canonical JSON array encoding of Terms.
// Output: `["commitment","tenancy","os","support_tier","upfront","payment_option"]`
// with no whitespace, matching json.dumps(..., separators=(",", ":")) on the Python side.
func CanonicalizeTerms(t Terms) string {
	// Use json.Marshal on a []string so Go's encoder emits the exact same
	// bytes as Python's json.dumps with compact separators: no spaces, fields
	// escaped identically.
	arr := [6]string{t.Commitment, t.Tenancy, t.OS, t.SupportTier, t.Upfront, t.PaymentOption}
	b, err := json.Marshal(arr)
	if err != nil {
		// []string marshalling cannot fail; panic preserves the invariant.
		panic("schema.CanonicalizeTerms: unreachable marshal error: " + err.Error())
	}
	return string(b)
}

// TermsHash returns the 128-bit hex digest (32 chars) of CanonicalizeTerms(t).
func TermsHash(t Terms) string {
	sum := sha256.Sum256([]byte(CanonicalizeTerms(t)))
	return hex.EncodeToString(sum[:16])
}
